### PR TITLE
openssl_csr_info: move main code to module_utils to allow easier implementation of diff mode

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Subset of #150.

##### ISSUE TYPE
- Refactoring Pull Request

##### COMPONENT NAME
openssl_csr_info

### CODE REVIEW

File: plugins/module_utils/crypto/module_backends/csr_info.py [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+                cryptography_oid_to_name(eku) for eku in ext_keyusage_ext.value
+            ]), ext_keyusage_ext.critical
+        except cryptography.x509.ExtensionNotFound:
+            return None, False
+
+    def _get_basic_constraints(self):
+        try:
+            ext_keyusage_ext = self.csr.extensions.get_extension_for_class(x509.BasicConstraints)
+            result = []
+            result.append('CA:{0}'.format('TRUE' if ext_keyusage_ext.value.ca else 'FALSE'))
```
Comment (Ajpantuso) [inline plugins/module_utils/crypto/module_backends/csr_info.py:247]:
```suggestion
            result['CA:{0}'.format('TRUE' if ext_keyusage_ext.value.ca else 'FALSE')]
```

File: plugins/module_utils/crypto/module_backends/csr_info.py [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+        except cryptography.x509.ExtensionNotFound:
+            return None, False
+
+    def _get_ocsp_must_staple(self):
+        try:
+            try:
+                # This only works with cryptography >= 2.1
+                tlsfeature_ext = self.csr.extensions.get_extension_for_class(x509.TLSFeature)
+                value = cryptography.x509.TLSFeatureType.status_request in tlsfeature_ext.value
+            except AttributeError as dummy:
```
Comment (Ajpantuso) [inline plugins/module_utils/crypto/module_backends/csr_info.py:260]:
```suggestion
            except AttributeError:
```

File: plugins/module_utils/crypto/module_backends/csr_info.py [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+        module.deprecate('The module is using the PyOpenSSL backend. This backend has been deprecated',
+                         version='2.0.0', collection_name='community.crypto')
+        return backend, CSRInfoRetrievalPyOpenSSL(module, content, validate_signature=validate_signature)
+    elif backend == 'cryptography':
+        if not CRYPTOGRAPHY_FOUND:
+            module.fail_json(msg=missing_required_lib('cryptography >= {0}'.format(MINIMAL_CRYPTOGRAPHY_VERSION)),
+                             exception=CRYPTOGRAPHY_IMP_ERR)
+        return backend, CSRInfoRetrievalCryptography(module, content, validate_signature=validate_signature)
+    else:
+        raise Exception('Unsupported value for backend: {0}'.format(backend))
```
Comment (Ajpantuso) [inline plugins/module_utils/crypto/module_backends/csr_info.py:462]:
```suggestion
        raise ValueError('Unsupported value for backend: {0}'.format(backend))
```

File: plugins/modules/openssl_csr_info.py [hunk 4]
```diff
@@ -664,53 +253,19 @@ def main():
...
-                getattr(crypto.X509Req, 'get_extensions')
-            except AttributeError:
-                module.fail_json(msg='You need to have PyOpenSSL>=0.15')
+    if module.params['content'] is not None:
+        data = module.params['content'].encode('utf-8')
+    else:
+        try:
+            with open(module.params['path'], 'rb') as f:
+                data = f.read()
+        except Exception as e:
```
Comment (Ajpantuso) [inline plugins/modules/openssl_csr_info.py:262]:
```suggestion
        except (IOError, OSError) as e:
```

File: plugins/module_utils/crypto/module_backends/csr_info.py [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+                aki = to_native(binascii.hexlify(aki))
+                aki = ':'.join([aki[i:i + 2] for i in range(0, len(aki), 2)])
+            result['authority_key_identifier'] = aki
+            result['authority_cert_issuer'] = aci
+            result['authority_cert_serial_number'] = acsn
+
+        result['extensions_by_oid'] = self._get_all_extensions()
+
+        result['signature_valid'] = self._is_signature_valid()
+        if self.validate_signature and not result['signature_valid']:
```
Comment (Ajpantuso) [inline plugins/module_utils/crypto/module_backends/csr_info.py:177]: This is for a new feature to be made available to users in a future PR?

**Review by Ajpantuso — CHANGES_REQUESTED** [review]
Similar to #203 just some minor cleanup items, but no functional differences from previous version.

File: plugins/module_utils/crypto/module_backends/csr_info.py (hunk 1 above)
Comment (felixfontein) [reply plugins/module_utils/crypto/module_backends/csr_info.py:247]: (I edited to insert ` = `)

File: plugins/module_utils/crypto/module_backends/csr_info.py (hunk 5 above)
Comment (felixfontein) [reply plugins/module_utils/crypto/module_backends/csr_info.py:177]: Yes, it's part of #150 (https://github.com/ansible-collections/community.crypto/pull/150/files#diff-16695b5a53f8b40bd252c59ca4278fc4096b92f07ae1f96edb890c508704bf76R192). The whole four PRs are part of that PR, I moved some of the commits to individual PRs to increase the chance that someone actually reviews them :)

File: plugins/module_utils/crypto/module_backends/csr_info.py (hunk 1 above)
Comment (Ajpantuso) [reply plugins/module_utils/crypto/module_backends/csr_info.py]: Thanks

File: plugins/module_utils/crypto/module_backends/csr_info.py (hunk 5 above)
Comment (Ajpantuso) [reply plugins/module_utils/crypto/module_backends/csr_info.py:177]: Cool, just checking that it was not pushed early or partially implemented.

**Review by Ajpantuso — APPROVED** [review]
LGTM, thanks!

**Comment by felixfontein** [comment]
@Ajpantuso thanks a lot for reviewing this one as well! :)

**Pull request merged**